### ISSUE TITLE
useLayoutEffect gets the initial state from useState instead of the current state when its input list never changes when an async method is fired

### ISSUE THREAD

**Issue by madou** [facebook/react#14713]
**Do you want to request a *feature* or report a *bug*?**

Report a bug.

**What is the current behavior?**

When using `useLayoutEffect` (having it fire only once during mount, and once during unmount), setting state outside of it and then having an async method run inside of it, results in it not being able to `useState` as expected. Instead of getting the latest value - it always gets the initial value.

Codesandbox here: https://codesandbox.io/s/9jm66px2z4
Actual usage here: https://github.com/madou/react-peer/blob/master/src/use-peer-state.tsx#L29

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem. Your bug will get fixed much faster if we can run your code and it doesn't have dependencies other than React. Paste the link to your JSFiddle (https://jsfiddle.net/Luktwrdm/) or CodeSandbox (https://codesandbox.io/s/new) example below:**

**What is the expected behavior?**

That it would get the current state when the async code is run. If I don't supply a input list to the hook it works as expected - however at the cost of repeated cleanup/initialisation.

I'm looking from the synonymous `componentDidMount()`/`componentWillUnmount()` hooks - was thinking using `useLayoutEffect` like so would do it.

Am I using it as expected? Is there a proper way to use it?

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**

Using `16.8.0-alpha.1`.

Cheers,

**Comment by madou** [comment]
Okay so using setState with the callback function it works as expected - cool! https://codesandbox.io/s/9yop1z283p

Makes sense - same pattern applies for class components as well.

I'm going to try and pin down the other problem I'm having where it's just not using the latest value. I'll report back soon.

**Comment by threepointone** [comment]
- this is as expected, when it first renders, the first function captures the value of number, and because the dependencies array never changes, it'll always use that value. as you noticed, the callback  version of `setNumber` fixes that 
- is there a specific reason you need to mutate the dom directly? you could instead write the value to another useState(), and use that as a prop to the div.
- it's not clear to me why you'd need `useLayoutEffect` here; why not `useEffect`?

**Comment by madou** [comment]
thanks for replying @threepointone 

honestly mutating the dom isn't the main point - please don't look at it much :) - if we talk about use cases ignoring setInterval and mutating the dom, I want to:

1. initialize a data provider in a hook once when the component mounts
2. cleanup/re-initialize a data provider if some props change
3. cleanup the data provider when component unmounts
4. access the latest available data in an async callback that is fired without user interaction (this step is breaking for me)

real world example is here https://github.com/madou/react-peer/blob/master/src/use-peer-state.tsx#L30

basically a client will connect, and I want to immediately send them _the latest data_, however they only ever see the initial value. am I using hooks in a silly way? 🤔 

with `useLayoutEffect` over `useEffect` - I don't have a strong reason to use it other than the docs mentioned its synonymous with `componentDidMount`/`componentWillUnmount` - is that still correct?

**Comment by gaearon** [comment]
If you need to access latest state at arbitrary times and don’t want to resubscribe when it changes, the simplest fix might be to commit latest state to a mutable ref:

https://reactjs.org/docs/hooks-faq.html#how-to-read-an-often-changing-value-from-usecallback

Another option is to have a layer on top of your connection code that makes resubscriptions cheap. Such as by deferring destroying the connection on next tick. Then use *that* layer from your effect and freely let it resubscribe when closed over state changes.

**Comment by gaearon** [comment]
>other than the docs mentioned its synonymous with componentDidMount/componentWillUnmount - is that still correct?

I think they also mentioned to not use it unless you actually have problems. :-) It makes your app slower. Just like cDM/cDU which block paint.

**Comment by madou** [comment]
thanks both - have some good ideas to think about. also need to remember hooks aren't magical and closure rules apply.. 😄
